docs/security/README.md

```diff
@@ -0,0 +1,244 @@
+
+Further any secret key - apart from the KMS-generated ones - are 256 bits long. The KMS-generated keys
+may be 256 bits but this depends on the KMS capabilities and configuration.
+
+The *Secure Channel* splits the object content into chunks of a fixed size of `65536` bytes. The last
+chunk may be smaller to avoid adding additional overhead and is treated specially to prevent truncation 
+attacks. The nonce value is 96 bits long and generated randomly per object / multi-part part. The 
+*Secure Channel* supports plaintexts up to `65536 * 2^32 = 256 TiB`. 
+
+#### Randomness
+
+The Minio server generates unique keys and other cryptographic values using a cryptographically
+secure pseudo-random number generator ([CSPRNG](#csprng)). However, in the context of SSE,
+the Minio server does not require that the CSPRNG generates values that are indistinguishable
+from truly random bit strings. Instead, it is sufficient if the generated values are unique - which
+is a weaker requirement. Nevertheless other parts - for example the TLS-stack - may require that
+CSPRNG-generated values are indistinguishable from truly random bit strings.
+
+### Server-Side Encryption with client-provided Keys
+
+SSE-C allows an S3 client to en/decrypt an object at the Minio server. Therefore the S3 client
+sends a secret key as part of the HTTP request. This secret key is **never** stored by the
+Minio server and only resides in RAM during the en/decryption process. 
+
+Minio does not assume or require that the client-provided key is unique. It may be used for
+multiple objects or buckets. Especially a single client-provided key may be used for all 
+objects - even though all objects must be treated as compromised if that key is ever compromised.
+
+#### Key rotation
+
+S3 clients can change the client-provided key of an existing object. Therefore an S3 client
+must perform a S3 COPY operation where the copy source and destination are equal. Further the
+COPY request headers must contain the current and the new client key:
+ - `X-Amz-Server-Side-Encryption-Customer-Key`: Base64 encoded new key.
+ - `X-Amz-Copy-Source-Server-Side-Encryption-Customer-Key`: Base64 encoded current key.
+
+Such a special COPY request is also known as S3 SSE-C key rotation.
+
+### Server-Side Encryption with a KMS
+
+SSE-S3 allows an S3 client to en/decrypt an object at the Minio server using a KMS. The Minio 
+server only assumes that the KMS provides two services:
+ 1. `GenerateKey`: Takes a key ID and generates a new data key from a master key referenced by
+                   the key ID. It returns the new data key in two different forms: The plain data key
+                   and the data key encrypted using the master key. 
+ 2. `DecryptKey`: Takes a key ID and an encrypted data key and returns the plain data key - the 
+                  decryption of the encrypted data key using the master key referenced by the key ID -
+                  on success or an error otherwise.
```

Comment:
the vault KMS does not return error for the DecryptKey API, the below sentence fragment can be removed.
```  
- on success or an error otherwise.
```

Reply:
It actually does. For example, if the master key does not exist or if the sent encrypted data key is not authentic. (Key decryption fails)

Reply:
oh yes, you are right